### PR TITLE
Fix immutable object virtual call getting no such method error

### LINKED ISSUE #43323: [Bug]: Generating a method call with an invalid method name when calling with a readonly object value
### Description

When trying to run nBallerina with the latest master (34a9f70) we get the fallowing error message.
```
error: No such method: $gen$(wso2&0047nballerina&0046comm&0046diagnostic&00580&0058$anonType$File$_0 & readonly)&0046filename
	at wso2.nballerina.comm.diagnostic.0:format(diagnostic.bal:123)
	   wso2.nballerina.comm.diagnostic.0.ConsolePrinter:print(printer.bal:18)
	   wso2.nballerina.0:main(main.bal:77)
```

This seems to be a regression introduced around the tag v2201.9.2 though this was not observed with the release version  9.2.

When diagnosing I observed in `diagnostic.class` file this,
## old (working)
Screenshot 2024-08-26 at 18 09 13

## broken (new)
Screenshot 2024-08-26 at 18 11 01

### Steps to Reproduce

1. Clone [nBallerina repository](https://github.com/ballerina-platform/nballerina/tree/main)
2. Run `bal build`
3. Go to `target\bin` and run `java -jar nballerina.jar ../../testSuite/01-boolean/not1-e.bal` (this should work with any test file with -e suffix)

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by HindujaB**
## Purpose
$subject
Fixes #43323 

## Approach
The fix added for https://github.com/ballerina-platform/ballerina-lang/issues/42988 missed a case where the attached function name generated for immutable object contains a module name prefix. 
A proper fix has to be added after fixing https://github.com/ballerina-platform/ballerina-lang/issues/43050. 

## Samples
```ballerina
public type File readonly & object {
    public function filename() returns string;
};

public function format(File file) {
    string line = file.filename();
}

public readonly class VirtualFile {
    *File;
    public function filename() returns string => "Hello";
}

public function main() {
    File d = new VirtualFile();
    format(d); // function name - (hinduja/test_listener:0:$anonType$File$_0 & readonly).filename
}
```

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmCodeGenUtil.java [hunk 1]
```diff
@@ -202,9 +202,15 @@ public static String rewriteVirtualCallTypeName(String value, BType objectType)
         objectType = getImpliedType(objectType);
         // The call name will be in the format of`objectTypeName.funcName` for attached functions of imported modules.
         // Therefore, We need to remove the type name.
+        Name originalName = objectType.tsymbol.originalName;
         if (!objectType.tsymbol.name.value.isEmpty() && value.startsWith(objectType.tsymbol.name.value)) {
```
Comment (ravinperera00) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmCodeGenUtil.java:206]: Would it improve readability if we extract out `objectType.tsymbol.name.value` and  `objectType.tsymbol.originalName` into variables as follows?
```
String typeName = objectType.tsymbol.name.value;
String originalName = objectType.tsymbol.originalName != null ? objectType.tsymbol.originalName.value : "";
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmCodeGenUtil.java [hunk 2]
```diff
@@ -202,9 +202,15 @@ public static String rewriteVirtualCallTypeName(String value, BType objectType)
...
             value = value.replace(objectType.tsymbol.name.value + ".", "").trim();
         }
+        // The call name will be in the format of`objectTypeOriginalName.funcName` for attached functions of
+        // object definitions. Therefore, We need to remove it.
+        if (originalName != null && !originalName.value.isEmpty() && value.startsWith(originalName.value)) {
```
Comment (ravinperera00) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmCodeGenUtil.java:211]: Will there be a case where reaching both the if conditions is important or can we change the second if condition to an else condition?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmCodeGenUtil.java [hunk 3]
```diff
@@ -200,10 +200,16 @@ public static String cleanupPathSeparators(String name) {
...
-            value = value.replace(objectType.tsymbol.name.value + ".", "").trim();
+        String typeName = objectType.tsymbol.name.value;
+        String originalName = objectType.tsymbol.originalName != null ? objectType.tsymbol.originalName.value : "";
+        if (!typeName.isEmpty() && value.startsWith(typeName)) {
+            // The call name will be in the format of`objectTypeName.funcName` for attached functions of imported
+            // modules. Therefore, We need to remove the type name.
+            value = value.replace(typeName + ".", "").trim();
+        } else if (value.startsWith(originalName)) {
+            // The call name will be in the format of`objectTypeOriginalName.funcName` for attached functions of
+            // object definitions. Therefore, We need to remove it.
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmCodeGenUtil.java:211]:
```suggestion
        String typeName = objectType.tsymbol.name.value;
        String originalName = objectType.tsymbol.originalName;
        if (!typeName.isEmpty() && value.startsWith(typeName)) {
            // The call name will be in the format of`objectTypeName.funcName` for attached functions of imported
            // modules. Therefore, We need to remove the type name.
            value = value.replace(typeName + ".", "").trim();
        } else if (originalName != null && value.startsWith(originalName)) {
            // The call name will be in the format of`objectTypeOriginalName.funcName` for attached functions of
            // object definitions. Therefore, We need to remove it.
```

Comment (warunalakshitha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmCodeGenUtil.java:211]: Will this work

**Review by warunalakshitha — APPROVED** [review]
LGTM

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmCodeGenUtil.java [hunk 4]
```diff
@@ -200,10 +200,16 @@ public static String cleanupPathSeparators(String name) {
 
     public static String rewriteVirtualCallTypeName(String value, BType objectType) {
         objectType = getImpliedType(objectType);
-        // The call name will be in the format of`objectTypeName.funcName` for attached functions of imported modules.
-        // Therefore, We need to remove the type name.
-        if (!objectType.tsymbol.name.value.isEmpty() && value.startsWith(objectType.tsymbol.name.value)) {
-            value = value.replace(objectType.tsymbol.name.value + ".", "").trim();
+        String typeName = objectType.tsymbol.name.value;
+        Name originalName = objectType.tsymbol.originalName;
+        if (!typeName.isEmpty() && value.startsWith(typeName)) {
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmCodeGenUtil.java:205]: Do we need this check typeName.isEmpty()? can typeName be null?

Comment (warunalakshitha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmCodeGenUtil.java:205]: Can we just use value.startsWith(typeName)

**Review by warunalakshitha — APPROVED** [review]

**Pull request merged**